pkg/authentication/middlewares.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2024 Canonical Ltd.
+// SPDX-License-Identifier: AGPL-3.0
+
+package authentication
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/http/types"
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+	"github.com/canonical/identity-platform-admin-ui/internal/tracing"
+)
+
+type Middleware struct {
+	allowListedEndpoints []string
+	tracer               tracing.TracingInterface
+	logger               logging.LoggerInterface
+	oauth2               OAuth2ContextInterface
+}
+
+func (m *Middleware) SetAllowListedEndpoints(endpointsPrefixes ...string) {
+	m.allowListedEndpoints = append(m.allowListedEndpoints, endpointsPrefixes...)
+}
+
+func (m *Middleware) isAllowListed(r *http.Request) bool {
+	endpoint := r.URL.Path
+	for _, prefix := range m.allowListedEndpoints {
+		if strings.HasPrefix(endpoint, prefix) {
+			return true
+		}
+	}
+	return false
+}
+
+func (m *Middleware) OAuth2Authentication(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx, span := m.tracer.Start(r.Context(), "authentication.Middleware.OAuth2Authentication")
+		defer span.End()
+		r = r.WithContext(ctx)
+
+		if m.isAllowListed(r) {
+			next.ServeHTTP(w, r.WithContext(ctx))
+			return
+		}
+
+		rawAccessToken, err := getBearerToken(r.Header)
+		if err != nil {
+			m.unauthorizedResponse(w, err)
+			return
+		}
```

Comment:
If that's the plan then we should talk mroe about this, since with the cookie based authentication we could rely on multiple pieces of info (id, access and refresh tokens), so we could (for users) refresh the access token if it's expired.
Let's talk about this and provide a plan for the future